dcplib/etl/__init__.py

```diff
@@ -0,0 +1,128 @@
+                            if not os.path.exists(f"{self.sd}/bundles/{bundle_uuid}.{bundle_version}/{f['name']}"):
+                                os.symlink(f"../../files/{f['uuid']}",
+                                           f"{self.sd}/bundles/{bundle_uuid}.{bundle_version}/{f['name']}")
+                            continue
+                except (FileNotFoundError, ):
+                    pass
+                files_to_fetch.append(f)
+            else:
+                logger.debug("Skipping non-metadata file %s/%s", bundle_uuid, f["name"])
+        return bundle_uuid, bundle_version, files_to_fetch
+
+    def enqueue_bundle_manifests(self, percent_complete, bundles_complete, total_bundles, f2f_futures, ff_futures,
+                                 executor):
+        logger.info("[%d%%] [%d/%d] Processing bundle batch",
+                    percent_complete, bundles_complete, total_bundles)
+        for future in concurrent.futures.as_completed(f2f_futures):
+            bundle_uuid, bundle_version, files_to_fetch = future.result()
+            if files_to_fetch:
+                logger.info("[%d%%] Fetching %d files for bundle %s @ %s",
+                            percent_complete, len(files_to_fetch), bundle_uuid, bundle_version)
+                for f in files_to_fetch:
+                    ff_futures.append(executor.submit(self.get_file, f, bundle_uuid, bundle_version))
+                    self.b2f[bundle_uuid + "." + bundle_version].add(f["name"])
+            else:
+                self.staged_bundles.append([bundle_uuid, bundle_version])
+        return len(f2f_futures)
+
+    def extract(self, query: dict = None, transformer: callable = None, loader: callable = None,
+                content_type_patterns=None, filename_patterns=None, max_workers=512):
+        if query is None:
+            query = self.default_bundle_query
+        if content_type_patterns is None:
+            content_type_patterns = self.default_content_type_patterns
+        os.makedirs("files", exist_ok=True)
+        os.makedirs("bundles", exist_ok=True)
+        f2f_futures, ff_futures = [], []
+        bundles_complete, percent_complete = 0, -1
+        with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
+            dss = hca.dss.DSSClient(swagger_url=f"{self.dss_host}/swagger.json")
+            total_bundles = dss.post_search(es_query=query, replica="aws")["total_hits"]
+            logging.info("Scanning %s bundles", total_bundles)
+            for b in dss.post_search.iterate(es_query=query, replica="aws", per_page=500):
+                bundle_uuid, bundle_version = b["bundle_fqid"].split(".", 1)
+                f2f_futures.append(executor.submit(self.get_files_to_fetch_for_bundle,
+                                                   bundle_uuid, bundle_version, content_type_patterns))
+
+                if len(f2f_futures) > max_workers * 2:
```

Comment:
Hm could you explain this clause?

Reply:
This is a heuristic to interleave fetching of bundle manifests with fetching of files from bundles. It serves to ask the API to check out files from bundles early and often, instead of waiting for all bundle manifests to be fetched before doing so. (It also serves to free up memory consumed by all the bundle manifest results.)